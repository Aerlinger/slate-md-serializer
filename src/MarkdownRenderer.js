import { Record } from 'immutable'

/**
 * String.
 */

const String = new Record({
  kind: 'string',
  text: ''
})

/**
 * A rule to (de)serialize text nodes. This is automatically added to the HTML
 * serializer so that users don't have to worry about text-level serialization.
 *
 * @type {Object}
 */

const TEXT_RULE = {
  serialize(obj, children) {
    if (obj.kind == 'string') {
      return children
    }
  }
}


/**
 * Markdown serializer.
 *
 * @type {Markdown}
 */

class Markdown {

  /**
   * Create a new serializer with `rules`.
   *
   * @param {Object} options
   *   @property {Array} rules
   * @return {Markdown} serializer
   */

  constructor(options = {}) {
    this.rules = [
      ...(options.rules || []),
      TEXT_RULE
    ];

    this.serializeNode = this.serializeNode.bind(this);
    this.serializeRange = this.serializeRange.bind(this);
    this.serializeString = this.serializeString.bind(this);
  }


  /**
   * Serialize a `state` object into an HTML string.
   *
   * @param {State} state
   * @return {String} markdown
   */

  serialize(state) {
    const { document } = state
    const elements = document.nodes.map(this.serializeNode)

    return elements.join('\n').trim();
  }

  /**
   * Serialize a `node`.
   *
   * @param {Node} node
   * @return {String}
   */

  serializeNode(node) {
    if (node.kind == 'text') {
      const ranges = node.getRanges()
      return ranges.map(this.serializeRange)
    }

    let children = node.nodes.map(this.serializeNode);
    children = children.flatten().length === 0 ? '' : children.flatten().join('')

    for (const rule of this.rules) {
      if (!rule.serialize) continue
      const ret = rule.serialize(node, children)
      if (ret) return ret
    }
  }

  /**
   * Serialize a `range`.
   *
   * @param {Range} range
   * @return {String}
   */

  serializeRange(range) {
    const string = new String({ text: range.text })
    const text = this.serializeString(string)

    return range.marks.reduce((children, mark) => {
      for (const rule of this.rules) {
        if (!rule.serialize) continue
        const ret = rule.serialize(mark, children)
        if (ret) return ret
      }
    }, text);
  }

  /**
   * Serialize a `string`.
   *
   * @param {String} string
   * @return {String}
   */

  serializeString(string) {
    for (const rule of this.rules) {
      if (!rule.serialize) continue
      const ret = rule.serialize(string, string.text)
      if (ret) return ret
    }
  }

}

/**
 * Export.
 */

export default Markdown
