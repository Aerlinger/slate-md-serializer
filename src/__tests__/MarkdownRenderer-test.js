import MarkdownRenderer from '../MarkdownRenderer'
const Markdown = new MarkdownRenderer();

test("parses heading1", () => {
  const output = Markdown.deserialize("# Heading", {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading2", () => {
  const output = Markdown.deserialize("## Heading", {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading3", () => {
  const output = Markdown.deserialize("### Heading", {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading4", () => {
  const output = Markdown.deserialize("#### Heading", {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading5", () => {
  const output = Markdown.deserialize("##### Heading", {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses heading6", () => {
  const output = Markdown.deserialize("###### Heading", {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses quote", () => {
  const text = `
> this is a quote
`
  const output = Markdown.deserialize(text, {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses quote with marks", () => {
  const text = `
> **bold** in a quote
`
  const output = Markdown.deserialize(text, {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses list items", () => {
  const text = `
 - one
 - two
`
  const output = Markdown.deserialize(text, {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});

test("parses list items with marks", () => {
  const text = `
 - one **bold**
 - *italic* two
`
  const output = Markdown.deserialize(text, {terse: true});
  expect(output.document.nodes).toMatchSnapshot();
});
