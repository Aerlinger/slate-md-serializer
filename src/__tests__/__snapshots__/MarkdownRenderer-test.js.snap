// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses heading1 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "1",
    "nodes": Array [
      Object {
        "characters": Array [
          Object {
            "marks": Array [],
            "text": "H",
          },
          Object {
            "marks": Array [],
            "text": "e",
          },
          Object {
            "marks": Array [],
            "text": "a",
          },
          Object {
            "marks": Array [],
            "text": "d",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "n",
          },
          Object {
            "marks": Array [],
            "text": "g",
          },
        ],
        "key": "0",
      },
    ],
    "type": "heading1",
  },
]
`;

exports[`parses heading2 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "4",
    "nodes": Array [
      Object {
        "characters": Array [
          Object {
            "marks": Array [],
            "text": "H",
          },
          Object {
            "marks": Array [],
            "text": "e",
          },
          Object {
            "marks": Array [],
            "text": "a",
          },
          Object {
            "marks": Array [],
            "text": "d",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "n",
          },
          Object {
            "marks": Array [],
            "text": "g",
          },
        ],
        "key": "3",
      },
    ],
    "type": "heading2",
  },
]
`;

exports[`parses heading3 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "7",
    "nodes": Array [
      Object {
        "characters": Array [
          Object {
            "marks": Array [],
            "text": "H",
          },
          Object {
            "marks": Array [],
            "text": "e",
          },
          Object {
            "marks": Array [],
            "text": "a",
          },
          Object {
            "marks": Array [],
            "text": "d",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "n",
          },
          Object {
            "marks": Array [],
            "text": "g",
          },
        ],
        "key": "6",
      },
    ],
    "type": "heading3",
  },
]
`;

exports[`parses heading4 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "10",
    "nodes": Array [
      Object {
        "characters": Array [
          Object {
            "marks": Array [],
            "text": "H",
          },
          Object {
            "marks": Array [],
            "text": "e",
          },
          Object {
            "marks": Array [],
            "text": "a",
          },
          Object {
            "marks": Array [],
            "text": "d",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "n",
          },
          Object {
            "marks": Array [],
            "text": "g",
          },
        ],
        "key": "9",
      },
    ],
    "type": "heading4",
  },
]
`;

exports[`parses heading5 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "13",
    "nodes": Array [
      Object {
        "characters": Array [
          Object {
            "marks": Array [],
            "text": "H",
          },
          Object {
            "marks": Array [],
            "text": "e",
          },
          Object {
            "marks": Array [],
            "text": "a",
          },
          Object {
            "marks": Array [],
            "text": "d",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "n",
          },
          Object {
            "marks": Array [],
            "text": "g",
          },
        ],
        "key": "12",
      },
    ],
    "type": "heading5",
  },
]
`;

exports[`parses heading6 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "16",
    "nodes": Array [
      Object {
        "characters": Array [
          Object {
            "marks": Array [],
            "text": "H",
          },
          Object {
            "marks": Array [],
            "text": "e",
          },
          Object {
            "marks": Array [],
            "text": "a",
          },
          Object {
            "marks": Array [],
            "text": "d",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "n",
          },
          Object {
            "marks": Array [],
            "text": "g",
          },
        ],
        "key": "15",
      },
    ],
    "type": "heading6",
  },
]
`;

exports[`parses list items 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "28",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "key": "25",
        "nodes": Array [
          Object {
            "characters": Array [
              Object {
                "marks": Array [],
                "text": "o",
              },
              Object {
                "marks": Array [],
                "text": "n",
              },
              Object {
                "marks": Array [],
                "text": "e",
              },
            ],
            "key": "24",
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "key": "27",
        "nodes": Array [
          Object {
            "characters": Array [
              Object {
                "marks": Array [],
                "text": "t",
              },
              Object {
                "marks": Array [],
                "text": "w",
              },
              Object {
                "marks": Array [],
                "text": "o",
              },
            ],
            "key": "26",
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
]
`;

exports[`parses list items with marks 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "34",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "key": "31",
        "nodes": Array [
          Object {
            "characters": Array [
              Object {
                "marks": Array [],
                "text": "o",
              },
              Object {
                "marks": Array [],
                "text": "n",
              },
              Object {
                "marks": Array [],
                "text": "e",
              },
              Object {
                "marks": Array [],
                "text": " ",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "bold",
                  },
                ],
                "text": "b",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "bold",
                  },
                ],
                "text": "o",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "bold",
                  },
                ],
                "text": "l",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "bold",
                  },
                ],
                "text": "d",
              },
            ],
            "key": "30",
          },
        ],
        "type": "list-item",
      },
      Object {
        "data": Object {},
        "isVoid": false,
        "key": "33",
        "nodes": Array [
          Object {
            "characters": Array [
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "italic",
                  },
                ],
                "text": "i",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "italic",
                  },
                ],
                "text": "t",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "italic",
                  },
                ],
                "text": "a",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "italic",
                  },
                ],
                "text": "l",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "italic",
                  },
                ],
                "text": "i",
              },
              Object {
                "marks": Array [
                  Object {
                    "data": Object {},
                    "type": "italic",
                  },
                ],
                "text": "c",
              },
              Object {
                "marks": Array [],
                "text": " ",
              },
              Object {
                "marks": Array [],
                "text": "t",
              },
              Object {
                "marks": Array [],
                "text": "w",
              },
              Object {
                "marks": Array [],
                "text": "o",
              },
            ],
            "key": "32",
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
]
`;

exports[`parses quote 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "19",
    "nodes": Array [
      Object {
        "characters": Array [
          Object {
            "marks": Array [],
            "text": "t",
          },
          Object {
            "marks": Array [],
            "text": "h",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "s",
          },
          Object {
            "marks": Array [],
            "text": " ",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "s",
          },
          Object {
            "marks": Array [],
            "text": " ",
          },
          Object {
            "marks": Array [],
            "text": "a",
          },
          Object {
            "marks": Array [],
            "text": " ",
          },
          Object {
            "marks": Array [],
            "text": "q",
          },
          Object {
            "marks": Array [],
            "text": "u",
          },
          Object {
            "marks": Array [],
            "text": "o",
          },
          Object {
            "marks": Array [],
            "text": "t",
          },
          Object {
            "marks": Array [],
            "text": "e",
          },
        ],
        "key": "18",
      },
    ],
    "type": "block-quote",
  },
]
`;

exports[`parses quote with marks 1`] = `
Array [
  Object {
    "data": Object {},
    "isVoid": false,
    "key": "22",
    "nodes": Array [
      Object {
        "characters": Array [
          Object {
            "marks": Array [
              Object {
                "data": Object {},
                "type": "bold",
              },
            ],
            "text": "b",
          },
          Object {
            "marks": Array [
              Object {
                "data": Object {},
                "type": "bold",
              },
            ],
            "text": "o",
          },
          Object {
            "marks": Array [
              Object {
                "data": Object {},
                "type": "bold",
              },
            ],
            "text": "l",
          },
          Object {
            "marks": Array [
              Object {
                "data": Object {},
                "type": "bold",
              },
            ],
            "text": "d",
          },
          Object {
            "marks": Array [],
            "text": " ",
          },
          Object {
            "marks": Array [],
            "text": "i",
          },
          Object {
            "marks": Array [],
            "text": "n",
          },
          Object {
            "marks": Array [],
            "text": " ",
          },
          Object {
            "marks": Array [],
            "text": "a",
          },
          Object {
            "marks": Array [],
            "text": " ",
          },
          Object {
            "marks": Array [],
            "text": "q",
          },
          Object {
            "marks": Array [],
            "text": "u",
          },
          Object {
            "marks": Array [],
            "text": "o",
          },
          Object {
            "marks": Array [],
            "text": "t",
          },
          Object {
            "marks": Array [],
            "text": "e",
          },
        ],
        "key": "21",
      },
    ],
    "type": "block-quote",
  },
]
`;
